crates/ruff_python_parser/src/semantic_errors.rs

```diff
@@ -313,7 +327,16 @@ impl SemanticSyntaxChecker {
         }
     }
 
-    pub fn visit_stmt<Ctx: SemanticSyntaxContext>(&mut self, stmt: &ast::Stmt, ctx: &Ctx) {
+    /// Check `stmt` for semantic syntax errors and update the checker's internal state.
+    ///
+    /// This should be followed by a call to [`SemanticSyntaxChecker::exit_stmt`] to reset any state
+    /// specific to scopes introduced by `stmt`, such as whether the body of a function is async.
+    pub fn enter_stmt<Ctx: SemanticSyntaxContext>(&mut self, stmt: &ast::Stmt, ctx: &Ctx) {
+        // check for errors
+        self.check_stmt(stmt, ctx);
+
+        self.save_checkpoint();
```

Comment:
Do we need to store the `checkpoint` for every statement or could we only store and restore it for some statements (e.g functions, lambdas, classes?)